document-generation/igce-document.ts

```diff
@@ -0,0 +1,270 @@
+    periodSheet.getCell("C3").value = fundingDocumentNumber;
+
+    // add each line item to the period sheet (top half)
+    const numberOfItems = periodLineItems.length;
+    const lineItemRows = periodSheet.getRows(periodSheetTopHalfInitialStartRow, numberOfItems);
+    lineItemRows?.forEach((row, index) => {
+      if (!uniqueIdiqClins.includes(periodLineItems[index].idiqClin)) {
+        uniqueIdiqClins.push(periodLineItems[index].idiqClin);
+      }
+
+      // fill each line item data
+      row.getCell("B").value = periodLineItems[index].clin; // 1000
+      row.getCell("C").value = periodLineItems[index].idiqClin; // "1000 Cloud";
+      row.getCell("D").value = periodLineItems[index].dowTaskNumber; // "4.2.2.1.1";
+      row.getCell("E").value = periodLineItems[index].serviceOffering; // "Compute";
+      row.getCell("F").value = periodLineItems[index].itemDescriptionOrConfigSummary; // "Compute - special project"
+      row.getCell("G").value = periodLineItems[index].monthlyPrice; // 159.03;
+      // periodLineItems[index].monthsInPeriod is calculated from period
+      row.getCell("H").value = convertPeriodToMonths(estimate.period); // 12;
+    });
+
+    // add additional rows to offset the numbers used on the summary sheet for idiq clins
+    const initialUsableRows = 12;
+    const lowerHalfEndingRow = periodSheetLowerHalfInitialStartRow + initialUsableRows;
+    const expectedLinesNeeded = summarySheetIdiqClinCounter + uniqueIdiqClins.length;
+    if (expectedLinesNeeded > initialUsableRows) {
+      const summarySheetAdditionalRows = uniqueIdiqClins.length + 2;
+      periodSheet.duplicateRow(lowerHalfEndingRow, summarySheetAdditionalRows, true);
+
+      // update the formula of the duplicated rows
+      periodSheet.eachRow((row, rowNumber) => {
+        updatePeriodFormulas(row, rowNumber, periodSheet);
+      });
+
+      // update formulas on summary sheet if additional rows were added
+      const mappingOfPeriodsToColumnOnSummarySheet: { [key: string]: string } = {
+        "Base Period": "C",
+        "Option Period 1": "D",
+        "Option Period 2": "E",
+        "Option Period 3": "F",
+        "Option Period 4": "G",
+        "Option Period 5": "H",
+        "Option Period 6": "I",
+      };
+      summarySheetColumnsToExtend.push(mappingOfPeriodsToColumnOnSummarySheet[periodSheet.name]);
+    }
+
+    // populate unique IDIQ CLIN items used on the period sheet (lower half)
+    const periodIdiqClinLines = periodSheet.getRows(
+      periodSheetLowerHalfInitialStartRow + summarySheetIdiqClinCounter,
+      uniqueIdiqClins.length
+    );
+    periodIdiqClinLines?.forEach((row, index) => {
+      row.getCell("C").value = uniqueIdiqClins[index];
+      allPeriodsIdiqClins.push(uniqueIdiqClins[index]);
+      summarySheetIdiqClinCounter++;
+    });
+  }
+
+  // populate period sheets
+  basePeriodLineItems.forEach(populatePeriodLineItems);
+  optionPeriodsLineItems.forEach(populatePeriodLineItems);
```

Comment:
I like it

Reply:
I wanted to break `populatePeriodLineItems` out on it's own, but it would require additional arguments and I don't think fits well with `forEach` 😢.